runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/DistributionData.java

```diff
@@ -29,36 +42,239 @@
  * the approximate value of those quantiles.
  */
-@AutoValue
-public abstract class DistributionData implements Serializable {
+public class DistributionData implements Serializable {
+  // Default list of percentiles and the corresponding extractors which should be sufficient for most cases.
+  private static final Map<Double, Function<DoublesSketch, Double>> DEFAULT_PERCENTILE_EXTRACTORS =
+      ImmutableMap.<Double, Function<DoublesSketch, Double>>builder()
+          .put(25.0D, new PercentileExtractor(25.0D))
+          .put(50.0D, new PercentileExtractor(50.0D))
+          .put(75.0D, new PercentileExtractor(75.0D))
+          .put(90.0D, new PercentileExtractor(90.0D))
+          .put(95.0D, new PercentileExtractor(95.0D))
+          .put(99.0D, new PercentileExtractor(99.0D))
+          .build();
+  // k = 256 should yield an approximate error ε of less than 1%
+  private static final int SKETCH_SUMMARY_SIZE = 256;
 
-  public abstract long sum();
+  private long sum;
+  private long count;
+  private long min;
+  private long max;
+  private Map<Double, Double> percentiles;
+  private transient UpdateDoublesSketch sketch;
 
-  public abstract long count();
+  /** Backward compatible static factory method. */
+  public static DistributionData empty() {
+    return new DistributionData();
+  }
 
-  public abstract long min();
+  /**
+   * Static factory method primary used for testing.
+   */
+  @VisibleForTesting
+  public static DistributionData create(long sum, long count, long min, long max) {
+    return create(sum, count, min, max, ImmutableMap.of());
+  }
 
-  public abstract long max();
+  /**
+   * Static factory method primary used for testing.
+   */
+  @VisibleForTesting
+  public static DistributionData create(
+      long sum, long count, long min, long max, Map<Double, Double> percentiles) {
+    return new DistributionData(sum, count, min, max, percentiles);
+  }
 
-  public static final DistributionData EMPTY = create(0, 0, Long.MAX_VALUE, Long.MIN_VALUE);
+  /**
+   * Constructor primarily used for testing. No reservoir is created, percentiles are directly
+   * assigned.
+   */
+  private DistributionData(
+      long sum, long count, long min, long max, Map<Double, Double> percentiles) {
+    this.sketch = null;
+    this.sum = sum;
+    this.count = count;
+    this.min = min;
+    this.max = max;
+    this.percentiles = percentiles;
+  }
 
-  public static DistributionData create(long sum, long count, long min, long max) {
-    return new AutoValue_DistributionData(sum, count, min, max);
+  private DistributionData() {
+    final DoublesSketchBuilder doublesSketchBuilder = new DoublesSketchBuilder();
+    this.sketch = doublesSketchBuilder.setK(SKETCH_SUMMARY_SIZE).build();
+    this.sum = 0L;
+    this.count = 0L;
+    this.min = Long.MAX_VALUE;
+    this.max = Long.MIN_VALUE;
+    this.percentiles = ImmutableMap.of();
```

Comment:
Let's have only one constructor that assigns value, other constructors just invokes another constructor.

Reply:
Fixed.